ansible_bender/db.py

```diff
@@ -83,16 +83,15 @@ def acquire(self):
         """
         while True:
             try:
-                with open(self._lock_path(), "r") as fd:
+                with os.fdopen(os.open(self._lock_path(), os.O_RDONLY ),'r') as fd:
                     # the file exists, ab changes the database
                     pid = fd.read()
                 logger.info("ab is running as PID %s", pid)
                 # logger.debug("stack trace: %s", traceback.extract_stack())
                 time.sleep(0.1)
             except FileNotFoundError:
                 # cool, let's take the lock
-                # FIXME: but this is not atomic, we should use open() for that
-                with open(self._lock_path(), "w") as fd:
+                with os.fdopen(os.open(self._lock_path(), os.O_WRONLY|os.O_EXCL|os.O_CREAT ),'w') as fd:
```

Comment:
control goes to the except block if the file does not exist, Now the file is created and PIDs will be written atomically using the os.O_EXCL|os.O_CREAT flags

Reply:
So it definitely won't be atomic - atomic means that in one operation you check for presence of the pid file and if it does not exist, you create it - acquite the lock. In this case, one operation is checking the presence and the other is to acquire the lock: this means there is a race condition when two threads could check the presence at the same time and then they would race who acquires it soon - that's exactly what we are trying to prevent.

I suggest reading more about the O_* flags and figure out what's the best practices of using them for the locking mechanism.